### PR TITLE
Fix/square brace range

### PR DESCRIPTION
**Pull request by jeswr**
Resolving https://github.com/fent/ret.js/issues/28.

### CODE REVIEW

File: lib/util.ts [hunk 1]
```diff
@@ -49,7 +49,8 @@ export const strToChars = (str: string): string => {
  */
 export const tokenizeClass = (str: string, regexpStr: string): [SetTokens, number] => {
   let tokens: SetTokens = [], rs: string[] | null, c: string;
-  const regexp = /\\(?:(w)|(d)|(s)|(W)|(D)|(S))|((?:(?:\\)(.)|([^\]\\]))-(?:\\)?([^\]]))|(\])|(?:\\)?([^])/g;
+  const regexp =
+  /\\(?:(w)|(d)|(s)|(W)|(D)|(S))|((?:(?:\\)(.)|([^\]\\]))-(((?:\\)])|(((?:\\)?([^\]])))))|(\])|(?:\\)?([^])/g;
```
Comment (fent) [inline lib/util.ts:53]: i regret parsing classes this way >_<

just a note, not a review of your changes. at some point, i want to refactor this function to tokenize character by character, not through a regex which can look confusing

File: test/main-test.js [hunk 2]
```diff
@@ -464,5 +464,150 @@ vows.describe('Regexp Tokenizer')
...
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.RANGE, from: 94, to: 57 },
+                ],
+              }],
+            });
+          },
+        },
```
Comment (fent) [inline test/main-test.js:553]: i get an error when trying to make a regex with this

    SyntaxError: Invalid regular expression: /[\^-9]/: Range out of order in character class

maybe this error check belongs in another PR, but i also don't think this test should suggest this is expected behavior

File: test/main-test.js [hunk 3]
```diff
@@ -464,5 +464,150 @@ vows.describe('Regexp Tokenizer')
...
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.RANGE, from: 91, to: 93 },
+                ],
+              }],
+            });
+          },
+        },
```
Comment (fent) [inline test/main-test.js:589]: maybe add closing square bracket that isn't escaped, as that is valid too. `[[-]]`

File: lib/util.ts [hunk 4]
```diff
@@ -61,9 +62,9 @@ export const tokenizeClass = (str: string, regexpStr: string): [SetTokens, numbe
       (rs[7] && {
         type: types.RANGE,
         from: (rs[8] || rs[9]).charCodeAt(0),
```
Comment (fent) [inline lib/util.ts:64]: does this need the last char from the captured group like `to` does?

Comment (jeswr) [reply lib/util.ts:64]: No, it is there in `to` because `rs[10]` will be `\\]` rather than `]` (similarly for other escaped chars), and so we need the code of the last char. By the looks `(rs[8] || rs[9])` has a similar effect in `from` because `rs[8]` does not include the `\\` if a character is escaped.

**Comment by fent** [comment]
thank you!
